### PR TITLE
Support Scala 2.13 in mlflow-spark

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/10505?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10505/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10505
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Update [mlflow-spark](https://github.com/databricks/mlflow-private/tree/6f0f0292278bc885b2b870f47500dd4fc46bf233/mlflow/java/spark) from Scala `2.12.6` to `2.13.10` A few notable things:
- Cannot update parent POM as mleap doesn't support scala 2.13 yet ([mleap/issues/811](https://github.com/combust/mleap/issues/811)) and mlflow-scoring depends on it.
- Updated Spark to 3.5.0 upon which DBR 14.x is built ([ref](https://docs.databricks.com/en/release-notes/runtime/index.html))
  - This requires a small change in autologging logic, as `DataSourceV2Relation` was removed from logical plan since 3.0.0 stable release. It was introduced in 3.0.0-preview (= current version we are using), but pulled back due to some issues. Hence the new logic fetches data source information from `LogicalRelation` node instead. The new logic is indeed just same as what we have for "REPL-ID aware" environment such as Databricks, so I assume the change is safe.
- While Spark 3.5 supports scala 2.13, we still need use scala 2.12 when using autologging from Python, as Pyspark is running on scala 2.12 anyway (before 4.0.0). As a workaround, we have to serve two jars - `mlflow-spark_2.12` and `mlflow-spark-2.13` meanwhile.
- Enabled cross-version tests (previously only enabled for 3.0.0-preview). 
  - Initially the tests were failed with Pyspark >= 3.2.0. Turned out it was due to "pin thread" mode that was defaulted since 3.2.0. As a workaround, setting env variable `PYSPARK_PIN_THREAD` to 'false'. See related code change for more details.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [x] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/java/pom.xml
Comment (B-Step62) [inline mlflow/java/pom.xml:148]: ref: https://www.slf4j.org/codes.html#StaticLoggerBinder

File: mlflow/java/spark/src/test/scala/org/mlflow/spark/autologging/SparkAutologgingIntegrationSuite.scala
Comment (B-Step62) [inline mlflow/java/spark/src/test/scala/org/mlflow/spark/autologging/SparkAutologgingIntegrationSuite.scala:216]: Somehow this broad match started to raise false-assert after version bump, even though the subscriber gets exact same calls and args. I guess it's just random ordering flakiness but can dig into more if this looks a concern.

File: mlflow/java/spark/pom.xml
Comment (WeichenXu123) [inline mlflow/java/spark/pom.xml:13]: I think we should still support scala 2.12, spark 3.5 still supports scala 2.12.

We can support to build mlflow-spark with either scala-2.12 or scala 2.13, like this project:

https://github.com/dmlc/xgboost/blob/3f4e22015a6796a0807650d7b2f97b108f8760df/jvm-packages/xgboost4j-spark/pom.xml#L12

and we can release 2 jars , one for scala 2.12, one for scala 2.13, like:
https://mvnrepository.com/artifact/ml.dmlc/xgboost4j-spark

WDYT ?

Comment (B-Step62) [reply mlflow/java/spark/pom.xml:13]: Yeah I faced exact issue when calling from Pyspark, and was working exactly same approach:) 
Will push another revision soon!

File: mlflow/java/spark/pom.xml
Comment (WeichenXu123) [inline mlflow/java/spark/pom.xml:3]: According to convention,
we should use format like `mlflow-spark_{scala.compat.version}`

File: mlflow/java/spark/pom.xml
Comment (WeichenXu123) [inline mlflow/java/spark/pom.xml:142]: Can we remove this ?

File: mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/DatasourceAttributeExtractor.scala
Comment (WeichenXu123) [inline mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/DatasourceAttributeExtractor.scala:65]: CC @cloud-fan to take a look.

File: mlflow/ml-package-versions.yml
Comment (WeichenXu123) [inline mlflow/ml-package-versions.yml:356]: I think we need to use `mvn install ...` if we want to run test in following script ?

File: mlflow/java/spark/pom.xml
Comment (B-Step62) [reply mlflow/java/spark/pom.xml:142]: Yup, but not super sure what it does?. I saw [your PR](https://github.com/mlflow/mlflow/pull/4181) to adding this profile, but would removing it change the process of publishing artifacts to the maven central?

Comment (WeichenXu123) [reply mlflow/java/spark/pom.xml:142]: It is for signing jars when publishing, we define `do-sign` profile in parent pom, so when running mvn with this profile enabled, the generated jars will be signed. But I think sub-project should inherit the `do-sign` profile for supporting signing.

File: mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/DatasourceAttributeExtractor.scala
Comment (cloud-fan) [inline mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/DatasourceAttributeExtractor.scala:59]: Spark can read from multiple paths in one shot. Do we want to display all of them or only the first path?

File: mlflow/_spark_autologging.py
Comment (B-Step62) [inline mlflow/_spark_autologging.py:246]: fyi: not changing behavior, just updated to the latest behavior (#4086)

File: tests/spark/autologging/conftest.py
Comment (B-Step62) [inline tests/spark/autologging/conftest.py:25]: Changing scope to function so as not to share same file paths across test, which might hide some failure.
For example, SparkAutologgingContext accumulates events to global variable `_table_info`, so those events can be carried over to the other test cases, unless cleaned explicitly (like I added in tear_down fixture). However, the unit tests have been passing so far because the file path in the event was same.

File: mlflow/java/spark/pom.xml
Comment (WeichenXu123) [inline mlflow/java/spark/pom.xml:15]: A reminder on follow-up task: we need to update MLflow release pipeline, to update stage of java releasing, so that it can release both scala 2.12 and 2.13 version

**Review by WeichenXu123 — APPROVED** [review]
LGTM, but let's waiting for @cloud-fan 's approval

**Review by cloud-fan — APPROVED** [review]
Spark part LGTM

File: mlflow/spark/__init__.py
Comment (harupy) [inline mlflow/spark/__init__.py:1118]: Comparing versions as a string can lead to errors. Can we use `packaging`?

Comment (B-Step62) [reply mlflow/spark/__init__.py:1118]: Sure! Fixing now.

File: tests/spark/autologging/conftest.py
Comment (harupy) [inline tests/spark/autologging/conftest.py:63]: Can we apply `@pytest.fixture(autouse=True)` on `clear_table_infos`?

File: tests/spark/autologging/utils.py
Comment (harupy) [inline tests/spark/autologging/utils.py:41]: Can we use the `monkeypatch` fixture instead of mutating `os.environ`?

File: mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala
Comment (harupy) [inline mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala:157]: /ai Is `format.forall(f => f == "text")` equivalent to `(format.isEmpty || format.get == "text")`?

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

File: tests/spark/autologging/conftest.py
Comment (B-Step62) [reply tests/spark/autologging/conftest.py:63]: Do you mean putting pytest annotation to the original def in `_spark_autologging.py`?

Comment (harupy) [reply tests/spark/autologging/conftest.py:63]: I see, `clear_table_infos` is imported from other module. Never mind.

File: mlflow/java/spark/pom.xml
Comment (xinzhao-db) [inline mlflow/java/spark/pom.xml:59]: why do we add these dependencies? It seems only mlflow/java/scoring are using them

File: mlflow/java/spark/pom.xml
Comment (xinzhao-db) [inline mlflow/java/spark/pom.xml:92]: and this one?

File: mlflow/java/spark/pom.xml
Comment (B-Step62) [reply mlflow/java/spark/pom.xml:59]: mlflow/java/spark uses them: https://github.com/mlflow/mlflow/blob/master/mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/ReplAwareSparkDataSourceListener.scala#L3-L6

**Pull request merged**
